### ISSUE TITLE
Auditing should support Java 8 Date & Time types [DATACMNS-411]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-data-commons#880]
**[Nick Williams](https://jira.spring.io/secure/ViewProfile.jspa?name=guitarking117)** opened **[DATACMNS-411](https://jira.spring.io/browse/DATACMNS-411?redirect=false)** and commented

This may or may not be in the right project. I think it is, but it may need to be moved to Spring Data Commons.

Currently, fields annotated `@CreatedDate` and `@LastModifiedDate` can be Joda Time's `DateTime` or (I think) `java.util.Date`, `java.util.Calendar`, or `java.sql.Timestamp`. If it doesn't support these last three types, it should.

Auditing should also support the Java 8 Date & Time API. Fields modified with one of these two annotations should be able to be `java.time.Instant`, `java.time.LocalDateTime`, `java.time.OffsetDateTime`, and `java.time.ZonedDateTime`

---

**Affects:** 1.7 M1 (Codd), 1.6.3 (Babbage SR2)

**Issue Links:**
- [SPR-11259](https://jira.spring.io/browse/SPR-11259) Add Converter implementations that convert legacy Date instances into JDK 8 date/time types
 (_**"depends on"**_)
- [DATAJPA-592](https://jira.spring.io/browse/DATAJPA-592) Auditing annotations not working when applied to methods

- [DATACMNS-307](https://jira.spring.io/browse/DATACMNS-307) Remove JodaTime dependency

**Referenced from:** commits https://github.com/spring-projects/spring-data-commons/commit/51b128fac1c17c8327d6fdd8c7ef87b039ac6c60

**Comment by adnklauser** [comment]
I think this should be re-opened. 

1. in 2021, support for something older than Java 8 might not be relevant anymore?
2. the workaround discussed above only works for JPA/Hibernate. In my case (azure-spring-data-cosmos), I can't work around it with a converter

**Comment by odrotbohm** [comment]
@adnklauser – Thanks for chiming in. Can you clarify what you like to see changed or added? Spring Data's auditing has supported Java 8 times for ages now. Looks like the ticket is a left over from times before it did so and just never got closed properly.

**Comment by adnklauser** [comment]
Hey @odrotbohm, 
thanks for chiming in so quickly :) 

It might actually be just a case of a misleading error message 😅. I am using the Azure cosmos DB spring data integration and I want to have java.time `OffsetDateTime` audit columns. When I create these, I get the error message from `org.springframework.data.auditing.DefaultAuditableBeanWrapperFactory#rejectUnsupportedType` (which doesn't list java.time in the list of supported types).

Upon closer examination of how `org.springframework.data.auditing.DefaultAuditableBeanWrapperFactory.DateConvertingAuditableBeanWrapper#getDateValueToSet` works, I realized that the default `DateTimeProvider` implementation provides a `LocalDateTime`, which cannot be automatically converted to an `OffsetDateTime`. 

Creating a custom `DateTimeProvider` that provides an `OffsetDateTime` in the first place, trivially solves the "conversion" issue in `getDateValueToSet`. From that point onward, serialization and deserialization of the java.time OffsetDateTime works perfectly fine, as you said.

**Comment by odrotbohm** [comment]
You'r observation is correct. We strongly believe in [time zones being a presentation layer problem](https://unix4lyfe.org/time/) and thus currently don't support any kinds of zoned date time types. Primarily because persisting them directly makes the values incomparable *unless* someone downstream unifies them to a reference time zone and persisting the zone separately. AFAIK, only Oracle currently supports persisting zoned times this way. That's why try to nudge folks into using `Instants` in their code, normalize to UTC and apply the potentially necessary time shift when preparing user facing representations.

Does that make sense?

**Comment by adnklauser** [comment]
Yes, fully agree. UTC `Instant` for point-in-time cases (apply time zone in presentation layer). LocalDateTime + separate time zone ID for "human" date+time (e.g., meeting at 9am).

And indeed, using `Instant` for the audit columns works out of the box 👍 Though I then find it strange that the default `org.springframework.data.auditing.CurrentDateTimeProvider` provides a `LocalDateTime`. _Technically_, it's not possible to convert a `LocalDateTime` into an `Instant` unless you assume a time zone (either system default or UTC). `org.springframework.data.convert.Jsr310Converters.LocalDateTimeToInstantConverter` silently assumes the system default time zone. 

In any case, my problem is solved (I have switched to `Instant`s as  they capture the nature of audit timestamps best). Thanks a lot for the discussion :)

**tl;dr for every one else stumbling across this ticket:** use `java.time.Instant` for the audit columns; not `java.time.OffsetDateTime`

**Comment by spyro2000** [comment]
I can't believe that this is still not working in 2021.

**Comment by gavenkoa** [comment]
> I can't believe that this is still not working in 2021...

Me too. @odrotbohm argument:

> We strongly believe in time zones being a presentation layer problem and thus currently don't support any kinds of zoned date time types. 

is very questionable if generalized (but the author pointed to `Instant`).  **In case of clock move backward one hour due to daylight saving we lose +/-1 hour in precision when using `LocalDateTime`. This is inaccessible for auditing framework.**

It was true that many DB were unable to store TZ with timestamp in the past. As I understand SQL 2003 defines `TIMESTAMP WITH TIMEZONE` and Oracle + Postgresql + H2 has such type.

JDBC 4.2 spec provides official mapping for `OffsetDateTime`. And `Instant` is not mentioned in tables B.4 & B.5 of the spec.

I assume Spring Data Auditing is just broken for the present.

`DateTimeProvider`  with `Optional.of(OffsetDateTime.now())` fixes the issue, but it shouldn't be an issue in the first place.

Prior discussions:

* https://github.com/spring-projects/spring-data-jpa/issues/1579 AuditAware not converting to ZonedDateTime anymore [DATAJPA-1242] 
* https://stackoverflow.com/questions/49170180/createdby-and-lastmodifieddate-are-no-longer-working-with-zoneddatetime
* https://github.com/spring-projects/spring-boot/issues/10743 Invalid date type exception when using JPA Auditor, SpringBoot 2.0.0.M5

I'll try to open a proper report justifying why `OffsetDateTime` should be supported.

**Comment by peterdieleman** [comment]
@spyro2000 And let me add, me too. Thanks for the snippet.

@gavenkoa Thanks, 100% your quote in bold part points out exactly why this is unacceptable for an auditing framework. This is as far as my internet queries got me. Did you try and open a report since?

**Comment by schauder** [comment]
Using `LocalDateTime` as the basis certainly was a mistake an we should probable fix that. 

That said, discussing this on an issue that was closed 8 years ago isn't helpful either.

For the record: The issue was closed because we do support Java 8 `java.time` classes to the extend that we actually want to: Anything with time zone is really out of place in my and not only in my opinion, especially since support for those in JPA and JDBC is ... sketchy.

In order to get this ticket here it's well earned rest: 
* If you want us to change the implementation basis for Auditing time stamps from `LocalDate` to `Instant` please create a ticket for it. For that one a PR even without a ticket has also a good chance of acceptance.
* If you think you really need support for `OffsetDateTime` you are free to create a fresh ticket as well. Just be aware that you'll probably have to do some convincing.

**Comment by gavenkoa** [comment]
@peterdieleman I forgot about this issue due to shift of priorities. Feel free to open a new one.

Key points:

* `Instant` is not in JDBC spec while `OffsetDateTime` is.
* Daylight saving breaks `LocalDateTime` very seriously.

**Comment by xenoterracide** [comment]
@schauder I want to use `OffsetDateTime`, it seems better due to the fact that you might be saving to a timestamp with timezone. I do believe hibernate has support too? maybe via `hibernate-java8`, I would be ok with having to add that dependency along with spring data jpa.

**Comment by xenoterracide** [comment]
currently I'm questioning the usability of this feature for this, better to write my own listener and only use it for user recording. Joda and Date/Calendar shouldn't be used anymore in my opinion

**Comment by xenoterracide** [comment]
(sorry can't edit comments) this issue should be reopened instead of a new one, IMO. As it hasn't actually been resolved and google will lead here.

**Comment by schauder** [comment]
@xenoterracide please open a new issue non the less. We use these issues to track changes and link them to releases. So reopening an issue that has released changes attached to it will break stuff.

If you reference this issue here there will be links between the issues, so people will be able to find any related issue.

**Comment by xenoterracide** [comment]
done
